macrosynergy/management/utils/sparse.py

```diff
@@ -0,0 +1,821 @@
+    p_eop = qdf_to_ticker_df(df, value_column="eop_lag")
+    if "grading" not in df.columns:
+        df["grading"] = np.nan
+    p_grading = qdf_to_ticker_df(df, value_column="grading")
+    assert set(p_value.columns) == set(p_eop.columns) == set(p_grading.columns)
+
+    # create dicts to store the dataframes and density stats
+    isc_dict: Dict[str, Any] = {}
+    density_stats: Dict[str, Dict[str, Any]] = {}
+    for ticker in p_value.columns:
+        df_temp, ddict = _get_diff_data(
+            ticker=ticker,
+            p_value=p_value,
+            p_eop=p_eop,
+            p_grading=p_grading,
+        )
+        isc_dict[ticker], density_stats[ticker] = df_temp, ddict
+
+    # flatten the density stats
+    _dstats_flat = [
+        (k, v["diff_density"], v["date_range"]) for k, v in density_stats.items()
+    ]
+
+    density_stats_df = (
+        pd.DataFrame(_dstats_flat, columns=["ticker", "changes_density", "date_range"])
+        .sort_values(by="changes_density", ascending=False)
+        .reset_index(drop=True)
+    )
+    if return_density_stats:
+        return isc_dict, density_stats_df
+    else:
+        return isc_dict
+
+
+class SubscriptableMeta(type):
+    """
+    Convenience metaclass to allow subscripting of methods on a class.
+    """
+
+    def __getitem__(cls, item):
+        if hasattr(cls, item) and callable(getattr(cls, item)):
+            return getattr(cls, item)
+        else:
+            raise KeyError(f"{item} is not a valid method name")
+
+
+class StandardDeviationMethods(metaclass=SubscriptableMeta):
+    """
+    Class to hold methods for calculating standard deviations.
+    Each method must comply to the following signature:
+        `func(s: pd.Series, **kwargs) -> pd.Series`
+
+    Currently supported methods are:
+
+    - `std`: Standard deviation
+    - `abs`: Absolute standard deviation
+    - `exp`: Exponentially weighted standard deviation
+    - `exp_abs`: Exponentially weighted absolute standard deviation
+    """
+
+    @staticmethod
+    def std(s: pd.Series, min_periods: int, **kwargs) -> pd.Series:
+        """
+        Calculate the standard deviation of a Series.
+
+        :param <pd.Series> s: The Series to calculate the standard deviation for.
+        :param <int> min_periods: The minimum number of periods required for the calculation.
+        :return: The standard deviation of the Series.
+        """
+        return s.expanding(min_periods=min_periods).std()
+
+    @staticmethod
+    def abs(s: pd.Series, min_periods: int, **kwargs) -> pd.Series:
+        """
+        Calculate the absolute standard deviation of a Series.
+
+        :param <pd.Series> s: The Series to calculate the absolute standard deviation for.
+        :param <int> min_periods: The minimum number of periods required for the calculation.
+        :return: The absolute standard deviation of the Series.
+        """
+        return s.abs().expanding(min_periods=min_periods).mean()
```

Comment:
Does this calculate the absolute standard deviation of a Series?

Reply:
@lsimonsen cleared up my confusion, but I suggest we go with `average absolute deviation` in the docstring